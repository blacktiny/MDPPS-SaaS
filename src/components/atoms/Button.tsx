import React from 'react'
import styled from '@emotion/styled'
import { css } from '@emotion/core'
import variables from '../../assets/styles/variables'
import rfs, { convertLetterSpacingToCssRem, boxModel, mq } from '../../utils/style'
import RButton, { ButtonProps as BaseButtonProps } from 'rsuite/lib/Button'

const { Fonts, Colors, DropShadow } = variables

const BaseStyle = css`
  border-radius: 4px;
  border: 3px solid transparent;
  font-weight: ${Fonts.Weight.Bold};
  line-height: ${rfs('22px')};
  transition: all 0.2s ease;
`

const PrimaryStyle = css`
  background-color: ${Colors.Blue[200]};
  box-shadow: ${DropShadow.standard};
  color: ${Colors.Gray[25]};
  font-size: ${rfs('16px')};
  letter-spacing: ${convertLetterSpacingToCssRem(22)};
  padding: ${boxModel('14px 40px')};

  &:hover {
    background-color: ${Colors.Blue[300]} !important;
  }
  &:focus {
    border-color: ${Colors.Blue[100]};
    background-color: ${Colors.Blue[200]} !important;
    outline: 0;
    border-radius: 7px;
  }
  &:focus&:active {
    box-shadow: none;
    background-color: ${Colors.Blue[400]} !important;
    border-radius: 5px;
  }
`

const SecondaryStyle = css`
  background-color: ${Colors.Gray[25]};
  border-color ${Colors.Gray[100]};
  color: ${Colors.Gray[400]};
  font-size: ${rfs('18px')};
  padding: ${boxModel('13.5px 40px')};

  ${mq({
    borderWidth: [`1px`, ``, ``, ``, `0.5px`]
  })}

  &:hover {
    background-color: ${Colors.Gray[50]};
  }
  &:focus {
    border-color: ${Colors.Blue[100]};
    background-color: ${Colors.Gray[25]} !important;
    outline: 0;
  }
  &:active {
    background-color: ${Colors.Gray[200]} !important;
    border: solid ${Colors.Gray[100]};
    color: ${Colors.Gray[25]} !important;
    padding: inherit;
  }
`

const StartStyle = css`
  align-self: flex-start;
`

const EndStyle = css`
  align-self: flex-end;
`

const SmallStyle = css`
  padding: ${boxModel('0.563rem 1.625rem')};
  &:focus {
    padding: ${boxModel('0.375rem 1.438rem')};
  }
  &:focus&:active {
    padding: ${boxModel('0.563rem 1.625rem')};
  }
`

const PairsStyle = css`
  padding-left: ${boxModel('26px')};
  padding-right: ${boxModel('26px')};
  font-size: ${rfs('16px')};
  &:focus {
    padding-left: ${boxModel('23px')};
    padding-right: ${boxModel('23px')};
  }
  &:focus&:active {
    padding-left: ${boxModel('26px')};
    padding-right: ${boxModel('26px')};
  }
`
const DisabledStyle = css`
  box-shadow: none;
  background-color: ${Colors.Gray[100]} !important;
  color: ${Colors.Gray[200]} !important;
  opacity: 1 !important;
  border: none;

  &:hover,
  &:focus {
    background-color: ${Colors.Gray[100]} !important;
  }
`

export const LinkButton = styled(RButton)`
  font-size: ${rfs('0.875rem')};
  font-size: ${rfs(Fonts.Size.XSmall)};
  line-height: 1.4;
  transition: all 0.5s ease-in-out;
  letter-spacing: 0.28px;
  padding: 0;
  align-self: flex-end;
  ${({ start }) => (start ? StartStyle : null)}
  ${({ end }) => (end ? EndStyle : null)}
  color: ${Colors.Blue[200]};
  a {
    color: ${Colors.Blue[200]};
  }
  &::first-letter {
    text-transform: capitalize;
  }
  ${({ small }) => (small ? SmallStyle : null)};
`

const BaseButton = styled(RButton)`
  ${BaseStyle};
  ${({ secondary }) => (secondary ? SecondaryStyle : PrimaryStyle)};
  ${({ small }) => (small ? SmallStyle : null)};
  ${({ pairs }) => (pairs ? PairsStyle : null)};

  ${({ disabled }) => (disabled ? DisabledStyle : null)}
  ${({ start }) => (start ? StartStyle : null)}
  ${({ end }) => (end ? EndStyle : null)}
`

export interface ButtonProps extends BaseButtonProps {
  link?: boolean
  primary?: boolean
  secondary?: boolean
  small?: boolean
  pairs?: boolean
  start?: boolean
  end?: boolean
  className?: string
}

const Button: React.FC<ButtonProps> = ({ link, children, secondary, pairs, end, start, ...props }) =>
  link ? (
    <LinkButton {...props} appearance="link" end={end ? end.toString() : null} start={start ? start.toString() : null}>
      {children}
    </LinkButton>
  ) : (
    <BaseButton
      {...props}
      secondary={secondary?.toString()}
      pairs={pairs?.toString()}
      end={end ? end.toString() : null}
      start={start ? start.toString() : null}
      appearance={secondary ? 'default' : 'primary'}
    >
      {children}
    </BaseButton>
  )

export default Button
