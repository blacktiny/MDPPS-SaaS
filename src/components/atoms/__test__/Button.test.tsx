import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import Button from '../Button'

const MyText = 'my test content'
test('should render children correctly', () => {
  const { queryByText } = render(<Button>{MyText}</Button>)

  expect(queryByText(MyText)).toBeInTheDocument()
})

test('should look as expected in the NORMAL state', () => {
  const { baseElement } = render(<Button>{MyText}</Button>)
  expect(baseElement).toMatchSnapshot()
})

test('should look as expected in the DISABLED state', () => {
  const { baseElement } = render(<Button disabled>{MyText}</Button>)

  expect(baseElement).toMatchSnapshot()
})

test('should look as expected in the NORMAL state with Secondary Attribute', () => {
  const { baseElement } = render(<Button secondary>{MyText}</Button>)

  expect(baseElement).toMatchSnapshot()
})

test('should look as expected in the DISABLED state with Secondary Attribute', () => {
  const { baseElement } = render(
    <Button secondary disabled={true}>
      {MyText}
    </Button>
  )

  expect(baseElement).toMatchSnapshot()
})

const clickCallback = jest.fn()
test('should trigger the click callback when clicked', () => {
  const { getByText } = render(
    <Button onClick={clickCallback}>{MyText}</Button>
  )
  fireEvent.click(getByText(MyText))
  expect(clickCallback).toHaveBeenCalledTimes(1)
})
