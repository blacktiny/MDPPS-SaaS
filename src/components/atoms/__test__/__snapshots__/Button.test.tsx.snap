// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should look as expected in the DISABLED state 1`] = `
.emotion-0 {
  border-radius: 4px;
  border: 3px solid transparent;
  font-weight: 700;
  line-height: calc(1.2625rem + 0.09375vw);
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #4284FC;
  box-shadow: 0 3px 6px rgba(0,0,0,0.1607843137254902);
  color: #FFFFFF;
  font-size: 1rem;
  -webkit-letter-spacing: 0.022rem;
  -moz-letter-spacing: 0.022rem;
  -ms-letter-spacing: 0.022rem;
  letter-spacing: 0.022rem;
  padding: 0.875rem 2.5rem;
  box-shadow: none;
  background-color: #F2F2F2 !important;
  color: #AAB8C2 !important;
  opacity: 1 !important;
  border: none;
}

.emotion-0:hover {
  background-color: #3469C7 !important;
}

.emotion-0:focus {
  border-color: #B6D0FD;
  background-color: #4284FC !important;
  outline: 0;
  border-radius: 7px;
}

.emotion-0:focus.emotion-0:active {
  box-shadow: none;
  background-color: #1E3E75 !important;
  border-radius: 5px;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: #F2F2F2 !important;
}

<body>
  <div>
    <button
      class="rs-btn rs-btn-primary emotion-0 rs-btn-disabled"
      disabled=""
      type="button"
    >
      my test content
      <span
        class="rs-ripple-pond"
      >
        <span
          class="rs-ripple"
        />
      </span>
    </button>
  </div>
</body>
`;

exports[`should look as expected in the DISABLED state with Secondary Attribute 1`] = `
.emotion-0 {
  border-radius: 4px;
  border: 3px solid transparent;
  font-weight: 700;
  line-height: calc(1.2625rem + 0.09375vw);
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #FFFFFF;
  border-color: #F2F2F2;
  color: #98a0ac;
  font-size: 1.125rem;
  padding: 0.84375rem 2.5rem;
  border-width: 1px;
  box-shadow: none;
  background-color: #F2F2F2 !important;
  color: #AAB8C2 !important;
  opacity: 1 !important;
  border: none;
}

@media (min-width:1290px) {
  .emotion-0 {
    border-width: 0.5px;
  }
}

.emotion-0:hover {
  background-color: #F9F9FA;
}

.emotion-0:focus {
  border-color: #B6D0FD;
  background-color: #FFFFFF !important;
  outline: 0;
}

.emotion-0:active {
  background-color: #AAB8C2 !important;
  border: solid #F2F2F2;
  color: #FFFFFF !important;
  padding: inherit;
}

.emotion-0:hover,
.emotion-0:focus {
  background-color: #F2F2F2 !important;
}

<body>
  <div>
    <button
      class="rs-btn rs-btn-default emotion-0 rs-btn-disabled"
      disabled=""
      secondary="true"
      type="button"
    >
      my test content
      <span
        class="rs-ripple-pond"
      >
        <span
          class="rs-ripple"
        />
      </span>
    </button>
  </div>
</body>
`;

exports[`should look as expected in the NORMAL state 1`] = `
.emotion-0 {
  border-radius: 4px;
  border: 3px solid transparent;
  font-weight: 700;
  line-height: calc(1.2625rem + 0.09375vw);
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #4284FC;
  box-shadow: 0 3px 6px rgba(0,0,0,0.1607843137254902);
  color: #FFFFFF;
  font-size: 1rem;
  -webkit-letter-spacing: 0.022rem;
  -moz-letter-spacing: 0.022rem;
  -ms-letter-spacing: 0.022rem;
  letter-spacing: 0.022rem;
  padding: 0.875rem 2.5rem;
}

.emotion-0:hover {
  background-color: #3469C7 !important;
}

.emotion-0:focus {
  border-color: #B6D0FD;
  background-color: #4284FC !important;
  outline: 0;
  border-radius: 7px;
}

.emotion-0:focus.emotion-0:active {
  box-shadow: none;
  background-color: #1E3E75 !important;
  border-radius: 5px;
}

<body>
  <div>
    <button
      class="rs-btn rs-btn-primary emotion-0"
      type="button"
    >
      my test content
      <span
        class="rs-ripple-pond"
      >
        <span
          class="rs-ripple"
        />
      </span>
    </button>
  </div>
</body>
`;

exports[`should look as expected in the NORMAL state with Secondary Attribute 1`] = `
.emotion-0 {
  border-radius: 4px;
  border: 3px solid transparent;
  font-weight: 700;
  line-height: calc(1.2625rem + 0.09375vw);
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #FFFFFF;
  border-color: #F2F2F2;
  color: #98a0ac;
  font-size: 1.125rem;
  padding: 0.84375rem 2.5rem;
  border-width: 1px;
}

@media (min-width:1290px) {
  .emotion-0 {
    border-width: 0.5px;
  }
}

.emotion-0:hover {
  background-color: #F9F9FA;
}

.emotion-0:focus {
  border-color: #B6D0FD;
  background-color: #FFFFFF !important;
  outline: 0;
}

.emotion-0:active {
  background-color: #AAB8C2 !important;
  border: solid #F2F2F2;
  color: #FFFFFF !important;
  padding: inherit;
}

<body>
  <div>
    <button
      class="rs-btn rs-btn-default emotion-0"
      secondary="true"
      type="button"
    >
      my test content
      <span
        class="rs-ripple-pond"
      >
        <span
          class="rs-ripple"
        />
      </span>
    </button>
  </div>
</body>
`;
